Add `@automattic/urls` package

oswian opened the pull request:
Resolves #88183

## Proposed Changes

Removes `lib/url/support.js` and migrates all URLs to a new `@automattic/urls` package.

## Testing Instructions

No functional changes, tests should pass 🟢 

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

oswian committed:
Add new package for Automattic URLs

chriskmnds reviewed (commented):
Thanks for working on this. It looks great! 🙌 

Just a couple of comments from me:

- I notice we are repeating some constants in `domains-table` package - in [url-support.ts](https://github.com/Automattic/wp-calypso/blob/trunk/packages/domains-table/src/utils/url-support.ts) file. Do you think we should import this package there and use these constants instead?
- Should we apply `localizeUrl` on the root? I notice some components apply the localization when using the URLs and some don't. I would think this is something to be applied always, as done in `domains-table`. Not sure if any caveats though. (and assuming we do apply it, I guess we ought to update the usages to not reapply)
- Not sure if something on my end, but my editor flags a few warnings like "cannot find module and type declarations" in [resolve-domain-status.tsx](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/resolve-domain-status.tsx) and [availability-messages.js](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/registration/availability-messages.js). Not sure if we are missing some setup there 🤔 

Perhaps CC @tyxla to give a confidence check?

tyxla commented on packages/urls/package.json:
```diff
@@ -0,0 +1,36 @@
+	"devDependencies": {
+		"@automattic/calypso-typescript-config": "workspace:^",
+		"typescript": "^5.3.3"
+	}
```
Should this package be `"private": true` for now? Are we aiming to publish it?

chriskmnds replied:
Thanks for the review! We do intend to publish and use this from NPM / import into other public package.

tyxla reviewed (approved):
Looks good to me! I'd only make the package private if we don't intend to publish it.

oswian commented:
Thanks for the review @chriskmnds & @tyxla

Addressing your comments Christos:

> I notice we are repeating some constants in domains-table package - in [url-support.ts](https://github.com/Automattic/wp-calypso/blob/trunk/packages/domains-table/src/utils/url-support.ts) file. Do you think we should import this package there and use these constants instead?

Yes lets avoid duplicates, updated: https://github.com/Automattic/wp-calypso/pull/88281/commits/c2084669b05d8b6f043b3ba55ccc6a57723bef7e

> Should we apply localizeUrl on the root? I notice some components apply the localization when using the URLs and some don't. I would think this is something to be applied always, as done in domains-table. Not sure if any caveats though. (and assuming we do apply it, I guess we ought to update the usages to not reapply)

Applying the localization on root sounds correct to me, however I do wonder why the eslint rule was explicitly [disabled](https://github.com/Automattic/wp-calypso/pull/80217/files#diff-0f0c7a88bd2750656ce4ff672f4f5434ba43a1d7aec70c4653ff270eda5e9e1eR1) in the first place. @lsl Any insight you can provide around why it was added, or why we might need to maintain it?

> Not sure if something on my end, but my editor flags a few warnings like "cannot find module and type declarations" in [resolve-domain-status.tsx](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/resolve-domain-status.tsx) and [availability-messages.js](https://github.com/Automattic/wp-calypso/blob/trunk/client/lib/domains/registration/availability-messages.js). Not sure if we are missing some setup there 🤔

Restarting `tsserver` in my editor usually resolve these warnings. For vscode: `cmd + shift + P -> Restart TS Server`. Let me know if that doesn't help.

lsl commented:
@lupus2k made that commit https://github.com/Automattic/wp-calypso/pull/80217/commits/218a973adf5ec2e884de1d047580ee0ac8f3bfad, but I'm guessing editing the file started triggering linter errors. That file likely predates the addition of the linter rule.

> Should we apply localizeUrl on the root?

`https://wordpress.com/support` should be safe to localize but in general I'd avoid this. `localizeUrl` is kind of temperamental about it's inputs, it expects full & english locale urls only.

> I notice we are repeating some constants in domains-table package - in [url-support.ts](https://github.com/Automattic/wp-calypso/blob/trunk/packages/domains-table/src/utils/url-support.ts) file. Do you think we should import this package there and use these constants instead?

How much are these constants actually being re-used? Unless there is a lot of re-use on these url's,  I would opt to delete them and hardcode the url's instead.

Unless these constants are used 100% of the time ([they're not](https://github.com/search?q=repo%3AAutomattic%2Fwp-calypso+wordpress.com%2Fsupport&type=code&p=3)), changing a url still requires you to grep around for other instances, and then you also have to hope there aren't any being obfuscated by `root` constants arbitrarily cutting off part way into the path.

chriskmnds commented on packages/domains-table/src/utils/resolve-domain-status.tsx:
```diff
@@ -1,4 +1,10 @@
+	INCOMING_DOMAIN_TRANSFER_STATUSES_IN_PROGRESS,
+	GDPR_POLICIES,
+	DOMAIN_EXPIRATION,
+} from '@automattic/urls';
```
Do we need to update the dependencies in `domains-table/package.json` for this?

oswian replied:
Thanks. Odd that this didn't trigger a build error 🤔 

Updated: [cc66333](https://github.com/Automattic/wp-calypso/pull/88281/commits/cc66333897b0fdbaab426e75271ca8be201a775f)

chriskmnds commented on packages/domains-table/src/utils/url-support.ts:
```diff
@@ -1,7 +0,0 @@
-import { localizeUrl } from '@automattic/i18n-utils';
-
-const root = localizeUrl( 'https://wordpress.com/support' );
```
If we are replacing this with the `@automattic/urls`, then we probably need to ensure either (1) `root` is localized in the same way, or (2) update the usages in `domains-table` to wrap with `localizeUrl`. I think let's probably go for the latter right now, as you'll otherwise may need to update everything else that imports from `@automattic/urls` to not wrap again (related to the second point in my previous comment).

oswian replied:
Good point.  It looks like instances within `domains-table` are already wrapped with `localizeUrl`, so with this change they are no longer 'double wrapped'. Also just an FYI (not suggesting this is ideal) but the [tests](https://github.com/Automattic/wp-calypso/commit/a5f70bbf80ac5a4ccb9340c61a3e6e586bc80fad#diff-ab021737268e6d80e413641836fea2bde886c52328613418b84f9b68ee8ec93dR72) indicate that 'double wrapping' is at least handled gracefully.

I've updated instances that were not using `localizeUrl`: [c742602](https://github.com/Automattic/wp-calypso/pull/88281/commits/c7426024f321be7a4284d8dc1aac4d6f01e62518)

chriskmnds replied:
thanks! all looks good :-)

chriskmnds reviewed (dismissed):
Thanks for the work again. The last change I believe needs a follow-up.

oswian commented:
Rebased and resolved some build errors.
